edge-util/test/Microsoft.Azure.Devices.Edge.Util.Test/OptionTest.cs

```diff
@@ -260,6 +279,41 @@ public async void TestForEachAsync()
             Assert.Equal(2, i);
         }
 
+        [Fact]
+        [Unit]
+        public async void TestForEachWithNoneMethodAsync()
+        {
+            var val = 0;
+            Option<int> some = Option.Some(3);
+            Option<int> none = Option.None<int>();
+            await some.ForEachAsync(
+            b =>
+            {
+                val += b;
+                return Task.CompletedTask;
+            },
+            () =>
+            {
+                val = -2;
+                return Task.CompletedTask;
+            });
+            Assert.Equal(3, val);
+
+            val = 0;
+            await none.ForEachAsync(
+            b =>
+            {
+                val += b;
+                return Task.CompletedTask;
+            },
+            () =>
+            {
+                val = -2;
+                return Task.CompletedTask;
+            });
+            Assert.Equal(-2, val);
+        }
+
```

Comment:
In theory the actions run on a different thread (e.g. val += b) that Assert.Equal() so the variables should be written/read using some synchronization mechanism (e.g volatile.read/write) Now it works only because 'await' optimizes and when it sees that the returned task is completed, it continues on the same thread. (also for the same reason, I'd put an 'await Task.Yield()' to the beginning of the actions to make it more realistic and test an actual async scenario.

Reply:
Viktor ran these tests a large number of times, and there were no failures, so I will not change.